### ISSUE TITLE
problem with some recent changes

### ISSUE THREAD

**Issue by RallyTronics** [alecthomas/entityx#143]
After grabbing the latest my code no longer compiles.  There appears to be a problem if a component contains a std::unique_ptr<>  I assume this is due to the create_from_copy additions but am not sure.  Any thoughts on how I can use the latest without losing the ability to use unique_ptr?

**Comment by zackthehuman** [comment]
Hmm... well, that makes sense since you can't copy or copy-construct a `std::unique_ptr`. There was a preexisting "copier" function, `assign_from_copy` that may also require that components be copy-constructable. 

If you checkout https://github.com/alecthomas/entityx/commit/249d97a975d0ef97be66e3ce201f169e3a27101c are you able to compile? This commit is before I introduced new things into entityx.

**Comment by RallyTronics** [comment]
Yeah, that works.  I guess I will have to lose the unique_ptrs.

On Fri, Apr 22, 2016 at 12:03 PM, Zack <notifications@github.com> wrote:

> Hmm... well, that makes sense since you can't copy or copy-construct a
> std::unique_ptr. There was a preexisting "copier" function,
> assign_from_copy that may also require that components be
> copy-constructable.
>
> If you checkout 249d97a
> <https://github.com/alecthomas/entityx/commit/249d97a975d0ef97be66e3ce201f169e3a27101c>
> are you able to compile? This commit is before I introduced new things into
> entityx.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/alecthomas/entityx/issues/143#issuecomment-213532007>
>



-- 
John Lagerquist
Chief Engineer
RallyTronics LLC
801-866-5981

**Comment by zackthehuman** [comment]
I will take a look at this when I have a few minutes. Obviously without a copy constructor, the default copy-constructor will try to copy a member that is a `std::unique_ptr` and it will fail. Another alternative is to create a copy constructor that sets that member to `nullptr` since it doesn't make sense to share ownership of uniques.

**Comment by RallyTronics** [comment]
I think your second solution may be the way to go, setting them to null in
the copy constructor.  I will give that a try.

On Fri, Apr 22, 2016 at 12:34 PM, Zack <notifications@github.com> wrote:

> I will take a look at this when I have a few minutes. Obviously without a
> copy constructor, the default copy-constructor will try to copy a member
> that is a std::unique_ptr and it will fail. Another alternative is to
> create a copy constructor that sets that member to nullptr since it
> doesn't make sense to share ownership of uniques.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/alecthomas/entityx/issues/143#issuecomment-213546035>
>



-- 
John Lagerquist
Chief Engineer
RallyTronics LLC
801-866-5981

**Comment by zackthehuman** [comment]
I'm interested in what you may be owning in a component via `unique_ptr`. Why not just have the resource be part of the component? Is it something like a texture?

**Comment by RallyTronics** [comment]
It points to a directed graph object I use to calculate network flow.  I am
not making a typical game :)  I will investigate as to whether or not I can
make it part of the component, it's been a while since I worked on this and
I can't remember why I went that route.  I would just use a standard
pointer but I am trying to maintain compatibility with the Cereal
serialization library.

On Fri, Apr 22, 2016 at 12:40 PM, Zack <notifications@github.com> wrote:

> I'm interested in what you may be owning in a component via unique_ptr.
> Why not just have the resource be part of the component? Is it something
> like a texture?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/alecthomas/entityx/issues/143#issuecomment-213547963>
>



-- 
John Lagerquist
Chief Engineer
RallyTronics LLC
801-866-5981
